File: admin-backend/admin-shared/admin-shared-persistence/src/main/java/io/github/faustofan/admin/shared/persistence/filter/TenantFilter.java
package io.github.faustofan.admin.shared.persistence.filter;

import org.babyfish.jimmer.sql.filter.Filter;
import org.babyfish.jimmer.sql.filter.FilterArgs;
import org.springframework.stereotype.Component;

import io.github.faustofan.admin.shared.common.constant.SystemConstants;
import io.github.faustofan.admin.shared.common.context.AppContext;
import io.github.faustofan.admin.shared.common.context.AppContextHolder;
import io.github.faustofan.admin.shared.common.exception.SystemException;
import io.github.faustofan.admin.shared.common.exception.errcode.SystemErrorCode;
import io.github.faustofan.admin.shared.persistence.model.TenantAwareProps;

/**
 * 租户过滤器
 */
@Component
public class TenantFilter implements Filter<TenantAwareProps> {
    /**
     * 添加租户过滤条件
     */
    @Override
    public void filter(FilterArgs<TenantAwareProps> args) {
        if (AppContextHolder.getContext() == null)
            throw new SystemException(SystemErrorCode.CONTEXT_MISSING);

        AppContext appContext = AppContextHolder.getContext();

        if (appContext.tenantId() == null)
            throw new SystemException(SystemErrorCode.TENANT_ID_MISSING);

        Long tenantId = appContext.tenantId();

        // 如果是系统操作， 比如用户第一次登录时， 就放行
        if (tenantId.equals(SystemConstants.SYSTEM_TENANT_ID))
            return;
        // 如果是超级管理员， 放行
        if (appContext.isSuperAdmin())
            return;
        // 其他情况， 都要加租户过滤
        args.where(args.getTable().tenantId().eq(tenantId));
    }
}

File: src/main/kotlin/io/github/faustofan/admin/config/jimmer/TenantFilter.kt
package io.github.faustofan.admin.config.jimmer

import io.github.faustofan.admin.common.context.AppContextHolder
import io.github.faustofan.admin.common.model.TenantAware
import io.github.faustofan.admin.common.model.tenantId
import org.babyfish.jimmer.sql.kt.ast.expression.eq
import org.babyfish.jimmer.sql.kt.filter.KFilter
import org.babyfish.jimmer.sql.kt.filter.KFilterArgs
import org.springframework.stereotype.Component

/**
 * 租户全局过滤器
 *
 * 自动为所有实现 TenantAware 接口的实体添加租户条件过滤
 * 确保数据隔离，用户只能访问本租户的数据
 *
 * 注意：超级管理员可以绕过此过滤器访问所有租户数据
 */
@Component
class TenantFilter : KFilter<TenantAware> {

    override fun filter(args: KFilterArgs<TenantAware>) {
        val context = AppContextHolder.get()

        // 超级管理员跳过租户过滤
        if (context.isSuperAdmin) {
            return
        }

        val tenantId = context.tenantId
        if (tenantId != null) {
            args.apply {
                where(table.tenantId eq tenantId)
            }
        }
    }
}

